README.md
# Lucifer Lightbringer

<img src="https://github.com/salkuadrat/lucifer/raw/master/lucifer.png" height="200" alt="Lucifer">

Lucifer is a fast, lightweight web framework in dart.

Built on top of native dart `HttpServer` to provide an elegant way to fulfill the needs of many modern web server these days.

Lucifer is open, efficient, and provide lots of features to handle dozen kinds of things.

## Installation 

[Install Dart SDK](https://dart.dev/get-dart)

You may start creating a new Lucifer project using lucy command.

```bash
pub global activate lucy

l create desire
```

The first command will activate command-line interface (CLI), named [Lucy](https://pub.dev/packages/lucy), to be accessible from your terminal. 

Then `l create desire` will generate your new project in the `desire` directory. 

Feel free to use any project name you want.

## Starting

Now we are ready to play with our web server.

You may open `main.dart` in your project `lib` directory to learn the structure of a simple lucifer application.

```dart
import 'package:lucifer/lucifer.dart';

void main() {
  final app = App();
  final port = env('PORT') ?? 3000;

  app.use(logger());

  app.get('/', (Req req, Res res) async {
    await res.send('Hello Detective');
  });

  await app.listen(port);
  print('Server running at http://${app.host}:${app.port}');

  app.checkRoutes();
}
```

You may test running it with the following command:

```bash
cd desire
l run
```

Now you may open `http://localhost:3000` in the web browser. 

If all went well, it will display `Hello Detective` and print the following message in your terminal.

```text
Server running at http://localhost:3000
```

## Fundamentals 

You may learn the fundamentals of Lucifer by understanding the code inside the `lib/main.dart` of your new project.

The short lines of code do several things behind the scene.

First, we import `lucifer` and create a web application by assigning a new `App` instance to `app`

```dart
import 'package:lucifer/lucifer.dart';
```

```dart
final app = App();
```

Then we set the server port with value `3000` from the `.env` file in your root project directory.

You may change it with any port you want.

```dart
final port = env('PORT') ?? 3000;
```

Next we tell it to listen to a GET request on root path `/` with `app.get()`

```dart
app.get('/', (Req req, Res res) async {
  
});
```

Every HTTP verbs comes with its own method in Lucifer: `get()`, `post()`, `put()`, `patch()`, `delete()`, with the first argument corresponds to the route path.

```dart
app.get('/', (Req req, Res res) {

});

app.post('/', (Req req, Res res) {

});

app.put('/', (Req req, Res res) {

});

app.patch('/', (Req req, Res res) {

});

app.delete('/', (Req req, Res res) {

});
```

For the second argument, you may see a callback function that will be called when an incoming request is processed, and send a response with it.

To handle the incoming request and send a response, you may write your code inside the callback function.

```dart
app.get('/', (Req req, Res res) async {
  await res.send('Hello Detective');
});
```

In the route callback function, Lucifer provides two objects, `req` and `res`, that represents `Req` and `Res` instance.

`Req` is a Request class built on top of native dart `HttpRequest`. 

It holds all information about the incoming request, such as request parameters, query string, headers, body, and more.

`Res` is a Response class built on top of native dart `HttpResponse`. 

It's mostly used to manipulate response and sending it to the client.

What you did before is sending a message string `Hello Detective` to the client using `res.send()`. This method sets the string in the response body, and then close the connection.

The last line of our code starts the server and listen for incoming requests on the specified `port`:

```dart
await app.listen(port);
print('Server running at http://${app.host}:${app.port}');
```

As an alternative, you may also use `app.listen()` like so:

```dart
// listen to the specified port and host
await app.listen(port, '127.0.0.1');

// listen to the specified port with callback
await app.listen(port, () {
  print('Server running at http://${app.host}:${app.port}');
});

// listen to the specified port and host with callback
await app.listen(port, 'localhost', () {
  print('Server running at http://${app.host}:${app.port}');
});
```

## Environment Variables

Environment is a set of variables known to a process (such as, ENV, PORT, etc). 

It's highly recommended to mimic production environment during development by reading it from `.env` file.

When we run `l create` command, a `.env` file is created in the root project directory, containing these values.

```text
ENV = development
PORT = 3000
```

Then you may access the `.env` value from your dart code using `env()` method:

```dart
void main() {
  final app = App();
  final port = env('PORT') ?? 3000; // get port from env

  // get ENV value to check if it's a development or production stage
  final environment = env('ENV'); 

  ...
}
```

For maximum security, we should always use environment variables for important things, such as database configurations and JSON Web Token (JWT) secret.

And one more thing... you may open `.gitignore` file in the root directory, and see that `.env` is included there. 

It means your `.env` file will not be uploaded to remote repository like GitHub, and your environment variables values will never be exposed to the public eyes.

## Request Parameters 

We've learned before that the `req` object holds the HTTP request informations. 

There are some properties of `req` that you will likely access in your application.

<table>
  <tr>
    <th>Property</th>
    <th>Description</th>
  </tr>
  <tr>
    <td>app</td>
    <td>holds reference to the Lucifer app object</td>
  </tr>
  <tr>
    <td>uriString</td>
    <td>URI string of the request</td>
  </tr>
  <tr>
    <td>path</td>
    <td>the URL path</td>
  </tr>
  <tr>
    <td>method</td>
    <td>the HTTP method being used</td>
  </tr>
  <tr>
    <td>params</td>
    <td>the route named parameters</td>
  </tr>
  <tr>
    <td>query</td>
    <td>a map object containing all the query string used in the request</td>
  </tr>
  <tr>
    <td>body</td>
    <td>contains the data submitted in the request body (must be parsed before you can access it)</td>
  </tr>
  <tr>
    <td>cookies</td>
    <td>contains the cookies sent by the request (needs the `cookieParser` middleware)</td>
  </tr>
  <tr>
    <td>protocol</td>
    <td>The request protocol (http or https)</td>
  </tr>
  <tr>
    <td>secure</td>
    <td>true if request is secure (using HTTPS)</td>
  </tr>
</table>

## GET Query String

Now you may learn how to retrieve the GET query parameters. 

Query string is the part that comes after URL path, and starts with a question mark `?` like `?username=lucifer`. 

Multiple query parameters can be added with character `&` like so:

```text
?username=lucifer&age=10000
```

How may we get the values?

Lucifer provides `req.query` object to make it easy to get all query values.

```dart
app.get('/', (Req req, Res res) {
  print(req.query);
});
```

The `req.query` object contains map of each query parameter. If there are no query, it will be an empty map or `{}`.

You may iterate on it with for loop. The following code will print each query key and its value:

```dart
for (var key in req.query.keys) {
  var value = req.query[key];
  print('Query $key: $value');
}
```

You may also access the individual value directly with `req.q()`

```dart
req.q('username'); // same as req.query['username']

req.q('age'); // same as req.query['age']
```

## POST Request Data

POST request data are sent by HTTP clients, such as from HTML form, or from a POST request sent using Postman or from an AJAX JavaScript code.

How may we access these data?

If the request data is sent as json with `Content-Type: application/json`, you may use `json()` middleware.

```dart
final app = App();

// use json middleware to parse json request body
// usually sent from REST API
app.use(json());
// use xssClean to clean the inputs
app.use(xssClean());
```

If it's sent as urlencoded `Content-Type: application/x-www-form-urlencoded`, you may use `urlencoded()` middleware.

```dart
final app = App();

// use urlencoded middleware to parse urlencoded request body
// usually sent from HTML form
app.use(urlencoded());
// use xssClean to clean the inputs
app.use(xssClean());
```

It all went well, you may access the parsed request data from `req.body`:

```dart
app.post('/login', (Req req, Res res) {
  final username = req.body['username'];
  final password = req.body['password'];
});
```

You may also use `req.data()` to access an individual request data directly:

```dart
app.post('/login', (Req req, Res res) {
  final username = req.data('username');
  final password = req.data('password');
});
```

Besides `json()` and `urlencoded()`, there are other available built in body parsers we can use. 

- `raw()` : to get request body as raw bytes
- `text()` : to get request body as a plain string
- `json()` : to parse json request body
- `urlencoded()` : to parse urlencoded request body
- `multipart()` : to parse multipart request body

To ensure the core framework stays lightweight, Lucifer will not assume anything about the request body. You may choose and apply the appropriate parser as needed in your application.

However, if you want to be safe and need to be able to handle all forms of request body, you may simply use the all-inclusive `bodyParser()` middleware.

```dart
final app = App();

app.use(bodyParser());
```

The `bodyParser` middleware will automatically detect the type of request body, and use the appropriate parser accordingly for each of incoming request in your application.

## Send Response 

In the example above, we have used `res.send()` to send a simple response to the client.

```dart
app.get('/', (Req req, Res res) async {
  await res.send('Hello Detective');
});
```

If you pass a string, lucifer will set `Content-Type` header to `text/html`.

If you pass a map or list object, it will set as `application/json`, and encode the data into JSON.

`res.send()` will set the correct `Content-Length` response header automatically.

`res.send()` also will close the connection when it's all done.

You may use `res.end()` method to send an empty response without any content in the response body.

```dart
app.get('/', (Req req, Res res) async {
  await res.end();
});
```

Another thing is you may also send the data directly without `res.send()` like so:

```dart
app.get('/string', (req, res) => 'string');

app.get('/int', (req, res) => 25);

app.get('/double', (req, res) => 3.14);

app.get('/json', (req, res) => { 'name': 'lucifer' });

app.get('/list', (req, res) => ['Lucifer',  'Detective']);
```

## HTTP Status Response

You may set the HTTP status response using `res.status()` method.

```dart
res.status(404).end();
```

or 

```dart
res.status(404).send('Not Found');
```

Or you may simply use `res.sendStatus()` for a shortcut.

```dart
// shortcut for res.status(200).send('OK');
res.sendStatus(200); 

// shortcut for res.status(403).send('Forbidden');
res.sendStatus(403);

// shortcut for res.status(404).send('Not Found');
res.sendStatus(404);

// shortcut for res.status(500).send('Internal Server Error');
res.sendStatus(500);
```

## JSON Response

Besides `res.send()` method, we may also use `res.json()` to send json data to the client. 

The method accepts a map or list object, and automatically encode it into json string with `jsonEncode()`

```dart
res.json({ 'name': 'Lucifer', 'age': 10000 });
```

```dart
res.json(['Lucifer', 'Detective', 'Amenadiel']);
```

## Cookies 

You may use `res.cookie()` to manage cookies in your application.

```dart
res.cookie('username', 'Lucifer');
```

The method accepts additional parameters with various options.

```dart
res.cookie(
  'username', 
  'Lucifer', 
  domain: '.luciferinheaven.com',
  path: '/admin',
  secure: true,
);
```

```dart
res.cookie(
  'username',
  'Lucifer',
  expires: Duration(milliseconds: DateTime.now().millisecondsSinceEpoch + 900000),
  httpOnly: true,
);
```

Here is some cookie parameters you may eat.

<table>
  <tr>
    <th>Value</th>
    <th>Type</th>
    <th>Description</th>
  </tr>
  <tr>
    <td>domain</td>
    <td>String</td>
    <td>Domain name for the cookie. Defaults to the domain name of the app</td>
  </tr>
  <tr>
    <td>expires</td>
    <td>Date</td>
    <td>Expiry date of the cookie in GMT. If not specified or set to 0, creates a session cookie that will be deleted when client close the browser.
</td>
  </tr>
  <tr>
    <td>httpOnly</td>
    <td>bool</td>
    <td>Flags the cookie to be accessible only by the web server</td>
  </tr>
  <tr>
    <td>maxAge</td>
    <td>int</td>
    <td>Convenient option for setting the expiry time relative to the current time in milliseconds</td>
  </tr>
  <tr>
    <td>path</td>
    <td>String</td>
    <td>Path for the cookie. Defaults to /</td>
  </tr>
  <tr>
    <td>secure</td>
    <td>bool</td>
    <td>Marks the cookie to be used with HTTPS only</td>
  </tr>
  <tr>
    <td>signed</td>
    <td>bool</td>
    <td>Indicates if the cookie should be signed</td>
  </tr>
  <tr>
    <td>sameSite</td>
    <td>bool or String</td>
    <td>Set the value of SameSite cookie</td>
  </tr>
</table>

A cookie can be deleted with

```dart
res.clearCookie('username');
```

Or you may use the following code to clear all cookies.

```dart
res.clearCookies();
```

## Secure Cookies

You may secure cookies in your application using `secureCookie()` middleware.

```dart
String cookieSecret = env('COOKIE_SECRET_KEY');

app.use(secureCookie(cookieSecret));
```

`COOKIE_SECRET_KEY` needs to be set in the `.env` file and should be a random string unique to your application.

## HTTP Headers

You may get the HTTP header of a request from `req.headers`

```dart
app.get('/', (req, res) {
  print(req.headers);
});
```

You may also use `req.get()` or `req.header()` to get an individual header value.

```dart
app.get('/', (req, res) {
  final userAgent = req.get('User-Agent');

  // same as 

  req.header('User-Agent');
});
```

To change the HTTP header of a response to client, you may use use `res.set()` and `res.header()`

```dart
res.set('Content-Type', 'text/html');

// same as 

res.header('Content-Type', 'text/html');
```

Here are some other ways to modify the Content-Type header of a response to the clinet.

```dart
res.type('.html'); // res.set('Content-Type', 'text/html');

res.type('html'); // res.set('Content-Type', 'text/html');

res.type('json'); // res.set('Content-Type', 'application/json');

res.type('application/json'); // res.set('Content-Type', 'application/json');

res.type('png'); // res.set('Content-Type', 'image/png');
```

## Redirects

Using redirects are common thing to do in a web application. 

You may redirect a response in your application with `res.redirect()` or `res.to()`

```dart
res.redirect('/get-over-here');

// same as 

res.to('/get-over-here');
```

It will create redirect with the default 302 status code.

You may also use it this way to set a custom status code.

```dart
res.redirect(301, '/get-over-here');

// same as 

res.to(301, '/get-over-here');
```

You may pass the path to `res.redirect()` with an absolute path (`/get-over-here`), an absolute URL (`https://scorpio.com/get-over-here`), a relative path (`get-over-here`), or `..` to go back one level.

```dart
res.redirect('../get-over-here');

res.redirect('..');
```

Or you may simply use `res.back()` to redirect back to the previous url, based on the HTTP Referer value sent by client in the request header (defaults to / if it's not set).

```dart
res.back();
```

## Routing 

Routing is the process of determining what should happen when a URL is called, and which parts of the application needs to handle the request.

In the example before we have used routing like so:

```dart
app.get('/', (req, res) async {

});
```

The code above creates a route that maps a root path `/` with HTTP GET method to the response we provide inside the callback function.

We may use named parameters to listen for custom request. 

Say we want to provide a profile API that accepts a string as username, and return the user details. 

However, we want the string parameter to be part of the URL, not as a query string. 

So we use the named parameters like so:

```dart
app.get('/profile/:username', (Req req, Res res) {
  // get username from URL parameter
  final username = req.params['username'];

  print(username);
});
```


CHANGELOG.md
## 1.0.0

- Initial release.

## 1.0.1 

- Update docs 

## 1.0.2 

- Update docs & description

## 1.0.3 

- Update dependencies

## 1.0.4 

- Clearing things

## 1.0.5

- Update info about Lucy CLI

## 1.0.6 

- Update Lucy CLI with L command

## 1.0.7

- Update Documentation 
- Update Lucy CLI `l m` command to generate a custom middleware
- Upgrade `cryptography` package dependency

lib/src/app.dart
import 'dart:async';
import 'dart:io';

import 'package:lucifer/lucifer.dart';
import 'package:queue/queue.dart';

import 'controller/controller.dart';
import 'handlers/handlers.dart';
import 'parsers/form_parser.dart';
import 'route/route.dart';
import 'route/route_matcher.dart';
import 'route/router.dart';
import 'exceptions.dart';
import 'local.dart';
import 'method.dart';
import 'request.dart';
import 'response.dart';
import 'socket.dart';
import 'typedef.dart';

///
/// Lucifer Core App
///
class App {
  App({
    this.idleTimeout,
    this.parallel = 50,
    this.shared = true,
    String? sslCertificate,
    String? privateKey,
    String? password,
  }) {
    // initialize
    _environment = env('ENV') ?? 'development';
    _requestQueue = Queue(parallel: parallel);
    _localData = LocalData();

    if (sslCertificate != null && privateKey != null && password != null) {
      String chain = Platform.script.resolve(sslCertificate).toFilePath();
      String key = Platform.script.resolve(privateKey).toFilePath();

      _securityContext = SecurityContext();
      _securityContext?.useCertificateChain(chain);
      _securityContext?.usePrivateKey(key, password: password);
    }
  }

  /// Gets or sets the timeout used for idle keep-alive connections.
  /// If no further request is seen within [idleTimeout] after the previous
  /// request was completed, the connection is dropped.
  ///
  /// Default is 120 seconds.
  ///
  /// Note that it may take up to 2 * idleTimeout before idle connection is aborted.
  /// To disable, set [idleTimeout] to null.
  final Duration? idleTimeout;

  /// The number of requests that can be processed at one time.
  ///
  /// Defaults to 50.
  ///
  final int parallel;

  /// If shared is true and more HttpServers from this isolate or other
  /// isolates are bound to the port, then the incoming connections will be
  /// distributed among all the bound HttpServers.
  ///
  /// Connections can be distributed over multiple isolates this way.
  ///
  /// Default shared is true.
  final bool shared;

  ///
  /// Environment status: development, testing or production
  ///
  late String _environment;

  String get environment => _environment;

  bool get isDevelopment => _environment == 'development';
  bool get isProduction => _environment == 'production';
  bool get isTesting => _environment == 'testing';

  ///
  /// Http Server
  ///
  late HttpServer _server;

  ///
  /// Local Data for this app
  ///
  late LocalData _localData;

  ///
  /// Requests queue
  ///
  late Queue _requestQueue;

  ///
  /// List of HTTP Route assigned for each [get,post,put,patch,delete] methods.
  ///
  List<Route> get routes => _routes;
  List<Route> _routes = [];

  void Function()? _onReady;

  ///
  /// To handle custom http errors
  ///
  final Map<int, Callback> _httpErrors = {};

  ///
  /// Security context for https
  ///
  SecurityContext? _securityContext;
  bool get _useSecure => _securityContext != null;

  ///
  /// the port used by this app
  ///
  int get port => _server.port;

  ///
  /// the host used by this app
  ///
  String get host => _server.address.host;

  ///
  /// Expose FormParser to be used in a route handler
  ///
  FormParser form() => FormParser();

  ///
  /// Create a socket server
  ///
  SocketServer socket(Req req, Res res) => SocketServer(req, res);

  /// Returns local data associated with the specified key
  ///
  /// If value is set, it will write to local data.
  ///
  T? local<T>(String key, [value]) {
    if (value == null) {
      return _localData.tryGet<T>(key);
    }

    _localData.set(key, value);
    return value;
  }

  ///
  /// Remove local data associated with the specified key
  ///
  T? removeLocal<T>(String key) {
    return _localData.remove(key);
  }

  ///
  /// Bind a custom error handler
  ///
  void on(int errorCode, Callback callback) {
    _httpErrors[errorCode] = callback;
  }

  ///
  /// Register route for all HTTP method
  ///
  App all(
    String path,
    dynamic middleware, [
    Callback? callback,
  ]) =>
      _addRoute(
        Method.all,
        path,
        middleware,
        callback,
      );

  ///
  /// Register a HEAD route
  ///
  App head(
    String path,
    dynamic middleware, [
    Callback? callback,
  ]) =>
      _addRoute(
        Method.head,
        path,
        middleware,
        callback,
      );

  ///
  /// Register a GET route
  ///
  App get(
    String path,
    dynamic middleware, [
    Callback? callback,
  ]) =>
      _addRoute(
        Method.get,
        path,
        middleware,
        callback,
      );

  ///
  /// Register a POST route
  ///
  App post(
    String path,
    dynamic middleware, [
    Callback? callback,
  ]) =>
      _addRoute(
        Method.post,
        path,
        middleware,
        callback,
      );

  ///
  /// Register a PUT route
  ///
  App put(
    String path,
    dynamic middleware, [
    Callback? callback,
  ]) =>
      _addRoute(
        Method.put,
        path,
        middleware,
        callback,
      );

  ///
  /// Register a DELETE route
  ///
  App delete(
    String path,
    dynamic middleware, [
    Callback? callback,
  ]) =>
      _addRoute(
        Method.delete,
        path,
        middleware,
        callback,
      );

  ///
  /// Register a PATCH route
  ///
  App patch(
    String path,
    dynamic middleware, [
    Callback? callback,
  ]) =>
      _addRoute(
        Method.patch,
        path,
        middleware,
        callback,
      );

  ///
  /// Register an OPTIONS route
  ///
  App options(
    String path,
    dynamic middleware, [
    Callback? callback,
  ]) =>
      _addRoute(
        Method.options,
        path,
        middleware,
        callback,
      );

  ///
  /// Register middleware(s) or apply a router
  ///
  /// We can use it like this.
  ///
  /// ```dart
  /// app.use(middleware); // single
  ///
  /// app.use([ middleware1, middleware2 ]); // multiple
  /// ```
  ///
  /// or with a specified path like this.
  ///
  /// ```dart
  /// app.use('/login', authMiddleware);
  ///
  /// app.use('/login', [ middleware1, middleware2 ]);
  /// ```
  ///
  /// To use it to apply a router.
  ///
  /// ```dart
  /// app.use(router); // apply the router for root
  /// ```
  ///
  /// or
  ///
  /// ```dart
  /// app.use('/api', router); // apply the router for /api
  /// ```
  ///
  App use(path, [middleware]) {
    bool isRouter = false;
    bool isMiddleware = false;

    // if only use 1 parameter (path)
    if (middleware == null) {
      // use for router, such as
      // app.use(router);
      if (path is Router) {
        isRouter = true;
      }

      // single middleware, such as
      // app.use(middleware);
      else if (path is Callback) {
        isMiddleware = true;
      }

      // multiple middlewares, such as
      // app.use([ middleware1, middleware2 ]);
      else if (path is List && path.isNotEmpty) {
        if (path.first is Callback) {
          isMiddleware = true;
        }
      }
    }

    // if use both parameter with correct path (string)
    if (path is String) {
      // use for router, such as
      // app.use('/api', router);
      if (middleware is Router) {
        isRouter = true;
      }

      // single middleware, such as
      // app.use('/api', middleware);
      else if (middleware is Callback) {
        isMiddleware = true;
      }

      // multiple middlewares, such as
      // app.use('/api', [ middleware1, middleware2 ]);
      else if (middleware is List && middleware.isNotEmpty) {
        if (middleware.first is Callback) {
          isMiddleware = true;
        }
      }
    }

    if (isRouter) {
      return _useRouter(path, middleware);
    }

    if (isMiddleware) {
      return _useMiddleware(path, middleware);
    }

    return this;
  }

  ///
  /// Register middleware(s).
  ///
  /// Use it like this.
  ///
  /// ```dart
  /// app.use(middleware);
  ///
  /// or
  ///
  /// app.use([ middleware1, middleware2 ]);
  /// ```
  ///
  /// or with a specified path like this.
  ///
  /// ```dart
  /// app.use('/login', authMiddleware);
  ///
  /// or
  ///
  /// app.use('/login', [ middleware1, middleware2 ]);
  /// ```
  ///
  App _useMiddleware(path, [middleware]) {
    List<Callback> middlewares = [];

    //
    // If only use 1 parameter (path), such as
    //
    // app.use(middleware);
    // app.use([ middleware1, middleware2 ]);
    //
    if (middleware == null) {
      // single middleware
      if (path is Callback) {
        middlewares = [path];
        path = '*';
      }

      // mutliple middlewares
      if (path is List && path.isNotEmpty) {
        if (path.first is Callback) {
          middlewares = [...path];
          path = '*';
        }
      }
    }

    // If both parameters are used, such as
    //
    // app.use('/api', middleware);
    // app.use('/api', [ middleware1, middleware2 ]);
    //
    if (path is String) {
      // single middleware
      if (middleware is Callback) {
        middlewares = [middleware];
      }

      // mutliple middlewares
      if (middleware is List && middleware.isNotEmpty) {
        if (middleware.first is Callback) {
          middlewares = [...middleware];
        }
      }
    }

    if (middlewares.isNotEmpty) {
      addRoute(Route(
        method: Method.all,
        path: path,
        middleware: middlewares,
      ));
    }

    return this;
  }

  ///
  /// Create Router to be used later with app.use()
  ///
  /// ```dart
  /// final router = app.router();
  ///
  /// router.get('/login', (req, res) {});
  /// router.get('/logout', (req, res) {});
  ///
  /// app.use('/auth', router);
  /// ```
  ///
  Router router([middleware]) {
    return Router(
      this,
      middleware: middleware,
      useRouteLater: true,
    );
  }

  ///
  /// Use router directly in a chained way.
  ///
  /// ```dart
  /// app.route('auth')
  ///   .get('/login', (req, res) {})
  ///   .get('/logout', (req, res) {});
  /// ```
  ///
  Router route(String path, [Controller? controller]) {
    Router router = Router(
      this,
      path: path,
      useRouteLater: false,
    );

    if (controller is Controller) {
      router.get('/', controller.index);
      router.post('/', controller.create);
      router.delete('/', controller.deleteAll);
      router.get('/:id', controller.view);
      router.put('/:id', controller.edit);
      router.delete('/:id', controller.delete);
    }

    return router;
  }

  ///
  /// Apply a defined router to this app.
  ///
  /// Use it like this.
  ///
  /// ```dart
  /// app.use(router); // apply the router for root /
  /// ```
  ///
  /// or
  ///
  /// ```dart
  /// app.use('/api', router); // apply the router for /api
  /// ```
  ///
  App _useRouter(path, [Router? router]) {
    // router without path
    // will apply to the root path
    if (path is Router && router == null) {
      path.apply(this);
    }

    // router to be applied to the specified path
    else if (path is String && router is Router) {
      router.apply(path);
    }

    return this;
  }

  ///
  /// Add and register a route.
  ///
  App _addRoute(
    Method method,
    String path,
    middleware, [
    Callback? callback,
  ]) {
    var or = middleware;
    var middlewares = <Callback>[];

    // route with a handler and middleware(s)
    if (callback != null) {
      // single middleware
      // such as: app.get('/', middleware, (req, res) {});
      if (or is Callback) {
        middlewares = [or];
      }

      // multiple middlewares
      // such as: app.get('/', [ middleware1, middleware2 ], (req, res) {});
      if (or is List && or.isNotEmpty) {
        if (or.first is Callback) {
          middlewares = [...or];
        }
      }
    }

    // route with a handler (and no middleware)
    // such as: app.get('/', (req, res) {});
    if (or is Callback && callback == null) {
      callback = or;
    }

    if (callback != null || middlewares.isNotEmpty) {
      addRoute(Route(
        path: path,
        method: method,
        middleware: middlewares,
        callback: callback,
      ));
    }

    return this;
  }

  ///
  /// Create a new route and add it to the routes list
  ///
  App addRoute(Route route) {
    _routes = [..._routes, route];
    return this;
  }

  App insertRoute(int index, Route route) {
    _routes.insert(index, route);
    return this;
  }

  ///
  /// Process incoming request
  ///
  Future<void> _processRequest(HttpRequest request) async {
    /// Initialize local data for this req res
    List initLocal = generateLocal();

    // Extract local data
    String localKey = initLocal[0];
    LocalData local = initLocal[1];

    local.set('___app___', this);

    Req req = Req(localKey, request);
    Res res = Res(localKey, request.response);

    res.local('___req___', req);

    // generate matched routes for this request
    final routes = RouteMatcher.match(
      method: req.httpMethod,
      path: req.path,
      routes: _routes,
    );

    try {
      // return 404 not found if no matched routes
      if (routes.isEmpty) {
        throw NotFoundException();
      }

      // to check if matched route contains callback with an empty function:
      //
      // app.get('/bloodyhell', (req, res) { });
      //
      // should return empty response, not return not found error.
      //
      bool hasCallback = false;

      // iterate through each matched routes
      for (final route in routes) {
        // parse params from its uri if the route has a handler
        // route can have null handler if it's a global middleware
        // registered in app.use
        req.params = RouteMatcher.params(route.path, req.uriString);

        if (route.hasCallback) {
          hasCallback = true;
          req.logs();
        }

        // iterate through each callback in the current route stack.
        for (final callback in route.stack) {
          // save result to be accessible in the next middleware or callback
          res.result = await callback(req, res);
          // res.log('${res.result}');

          // check return value, if false, close connection.
          if (res.result == false) {
            return await res.close();
          }
        }

        // only process result from route callback, not middleware.
        //
        // have to check hasCallback first, since route created from app.use()
        // can contains only middleware and null callback.
        if (route.hasCallback) {
          // if the returned data from callback saved at `res.result` is
          // not null and not boolean (true/false),
          // and res is still open
          if (res.open && res.hasValidResult) {
            await _handleResponse(req, res, res.result);
          }
        }
      }

      // if matched route has callback with an empty function,
      // it needs to return an empty response to the client.
      //
      // app.get('/bloodyhell', (req, res) { });
      //
      if (res.open && hasCallback) {
        return await res.end();
      }

      // return 404, if somehow still not closed
      // in the end of the process
      if (res.open) {
        throw NotFoundException();
      }
    } on HttpServerException catch (e) {
      if (e.stackTrace != null) print(e.stackTrace);
      await _handleHttpException(req, res, e.status, e.message);
    } catch (e, stackTrace) {
      print(stackTrace);
      String message = isProduction ? getStatusMessage(500) : e.toString();
      await _handleHttpException(req, res, 500, message);
    } finally {
      removeLocal(localKey);
      res.close();
    }
  }

  ///
  /// Handles Http Exception
  ///
  FutureOr _handleHttpException(
    Req req,
    Res res,
    int status,
    String message,
  ) async {
    req.logs();

    if (_httpErrors.containsKey(status)) {
      final callback = _httpErrors[status];

      if (callback is Function) {
        final result = await callback?.call(req, res);

        if (res.open) {
          if (result == false) {
            return await res.close();
          }

          await _handleResponse(req, res, result);
        }
      }
    }

    if (res.open) {
      await res.status(status).send(message);
    }
  }

  ///
  /// Handles a response according to data type
  ///
  FutureOr _handleResponse(Req req, Res res, data) async {
    if (data != null && data is! bool) {
      for (DataHandler handler in handlers) {
        if (handler.canHandle(data)) {
          await handler.process(req, res, data);
          return await res.close();
        }
      }
    }
  }

  ///
  /// Start the server by listening to the specified `host` and `port`.
  ///
  /// Can be called in two ways.
  ///
  /// ```dart
  /// await app.listen(port, () { });
  ///
  /// await app.listen(port, host, () { });
  ///
  /// ```
  ///
  /// Default host is localhost and port is 3000.
  ///
  Future<void> listen(int port, [host, void Function()? onReady]) async {
    // default host and port
    String defaultHost = 'localhost';

    // app.listen(3000);
    if (host == null && onReady == null) {
      host = defaultHost;
    }
    // app.listen(3000, () { });
    else if (host is void Function() && onReady == null) {
      onReady = host;
      host = defaultHost;
    }

    assert(port is int && host is String, 'Invalid parameter');

    if (_useSecure) {
      // bind https server to specified port
      _server = await HttpServer.bindSecure(
        host,
        port,
        _securityContext!,
        shared: shared,
      );
    } else {
      // bind server to specified port
      _server = await HttpServer.bind(
        host,
        port,
        shared: shared,
      );
    }

    // set custom idleTimeout
    if (idleTimeout is Duration) {
      _server.idleTimeout = idleTimeout;
    }

    // start listening for incoming requests
    _server.listen((HttpRequest request) {
      // add each request to queue for processing
      _requestQueue.add(() async {
        await _processRequest(request);
      });
    });

    // save onReady for restart
    _onReady = onReady;

    // inform server is ready
    onReady?.call();
  }

  /// Close the server.
  ///
  /// Default `force` is false.
  ///
  Future<void> close({bool force = false}) async {
    await _server.close(force: force);
  }

  ///
  /// Restart the server
  ///
  Future<void> restart() async {
    final port = this.port;
    final host = this.host;
    final onReady = _onReady;

    await close(force: true);
    await listen(port, host, onReady);
  }

  ///
  /// Print out registered routes to the console.
  ///
  /// Helpful to inspect what's available.
  ///
  void checkRoutes() {
    print('');
    for (var route in _routes) {
      if (route.hasCallback) {
        print(route.toString());
      }
    }
    print('');
  }
}
